File: frontend_gn/src/constants/permissions.js
/**
 * Constantes de Permissions
 * Définit toutes les permissions disponibles dans le système
 * À synchroniser avec backend_gn/Utilisateur/permissions.py
 */

export const PERMISSIONS = {
  // Dashboard
  DASHBOARD_VIEW: 'dashboard.view',
  
  // Fiches Criminelles
  FICHES_VIEW: 'fiches.view',
  FICHES_CREATE: 'fiches.create',
  FICHES_EDIT: 'fiches.edit',
  FICHES_DELETE: 'fiches.delete',
  
  // Biométrie
  BIOMETRIE_VIEW: 'biometrie.view',
  BIOMETRIE_ADD: 'biometrie.add',
  BIOMETRIE_EDIT: 'biometrie.edit',
  BIOMETRIE_DELETE: 'biometrie.delete',
  
  // Rapports
  REPORTS_VIEW: 'reports.view',
  REPORTS_CREATE: 'reports.create',
  REPORTS_EXPORT: 'reports.export',
  
  // Notifications & Emails
  NOTIFICATIONS_VIEW: 'notifications.view',
  
  // Intelligence Artificielle
  IA_VIEW_RESULTS: 'ia.view_results',
  IA_USE_RECOGNITION: 'ia.use_recognition',
  IA_USE_PREDICTION: 'ia.use_prediction',
  
  // Audit
  AUDIT_VIEW: 'audit.view',
  AUDIT_VIEW_OWN: 'audit.view_own',
  AUDIT_VIEW_ALL: 'audit.view_all',
  
  // Utilisateurs
  USERS_VIEW: 'users.view',
  USERS_CREATE: 'users.create',
  USERS_EDIT: 'users.edit',
  USERS_DELETE: 'users.delete',
  
  // Rôles et Permissions
  ROLES_VIEW: 'roles.view',
  ROLES_MANAGE: 'roles.manage',
  
  // Enquêtes
  INVESTIGATIONS_CREATE: 'investigations.create',
  INVESTIGATIONS_EDIT: 'investigations.edit',
  INVESTIGATIONS_CLOSE: 'investigations.close',
  INVESTIGATIONS_VIEW: 'investigations.view',
  
  // Suspects
  SUSPECTS_CREATE: 'suspects.create',
  SUSPECTS_EDIT: 'suspects.edit',
  SUSPECTS_VIEW: 'suspects.view',
  
  // Preuves
  EVIDENCE_MANAGE: 'evidence.manage',
  EVIDENCE_VIEW: 'evidence.view',
  
  // Analyses
  ANALYTICS_VIEW: 'analytics.view',
}

/**
 * Configuration des permissions par rôle
 * Correspond exactement aux spécifications
 */
export const ROLE_PERMISSIONS = {
  'Administrateur Système': {
    description: 'Accès complet à toutes les fonctionnalités du système',
    color: 'red',
    permissions: Object.values(PERMISSIONS), // Toutes les permissions
  },
  
  'Observateur': {
    description: 'Consultation uniquement, sans modification',
    color: 'gray',
    permissions: [
      PERMISSIONS.DASHBOARD_VIEW,
      PERMISSIONS.FICHES_VIEW,
      PERMISSIONS.BIOMETRIE_VIEW,
      PERMISSIONS.REPORTS_VIEW,
      PERMISSIONS.REPORTS_EXPORT,
      PERMISSIONS.NOTIFICATIONS_VIEW,
      PERMISSIONS.IA_VIEW_RESULTS,
      PERMISSIONS.INVESTIGATIONS_VIEW,
      PERMISSIONS.SUSPECTS_VIEW,
      PERMISSIONS.EVIDENCE_VIEW,
    ],
  },
  
  'Enquêteur Principal': {
    description: 'Responsable des enquêtes, création et gestion complète des dossiers criminels',
    color: 'blue',
    permissions: [
      //  Tableau de bord
      PERMISSIONS.DASHBOARD_VIEW,
      
      //  Fichier criminel (création, modification, suppression)
      PERMISSIONS.FICHES_VIEW,
      PERMISSIONS.FICHES_CREATE,
      PERMISSIONS.FICHES_EDIT,
      PERMISSIONS.FICHES_DELETE,
      
      //  Biométrie (ajout de données faciales, empreintes)
      PERMISSIONS.BIOMETRIE_VIEW,
      PERMISSIONS.BIOMETRIE_ADD,
      PERMISSIONS.BIOMETRIE_EDIT,
      
      //  Rapport (génération et consultation)
      PERMISSIONS.REPORTS_VIEW,
      PERMISSIONS.REPORTS_CREATE,
      PERMISSIONS.REPORTS_EXPORT,
      
      //  Intelligence artificielle (reconnaissance faciale, prédiction)
      PERMISSIONS.IA_VIEW_RESULTS,
      PERMISSIONS.IA_USE_RECOGNITION,
      PERMISSIONS.IA_USE_PREDICTION,
      
      // Enquêtes
      PERMISSIONS.INVESTIGATIONS_CREATE,
      PERMISSIONS.INVESTIGATIONS_EDIT,
      PERMISSIONS.INVESTIGATIONS_CLOSE,
      PERMISSIONS.INVESTIGATIONS_VIEW,
      
      // Suspects
      PERMISSIONS.SUSPECTS_CREATE,
      PERMISSIONS.SUSPECTS_EDIT,
      PERMISSIONS.SUSPECTS_VIEW,
      
      // Preuves
      PERMISSIONS.EVIDENCE_MANAGE,
      PERMISSIONS.EVIDENCE_VIEW,
      
      // Analyses
      PERMISSIONS.ANALYTICS_VIEW,
      PERMISSIONS.NOTIFICATIONS_VIEW,
    ],
  },
  
  'Analyste': {
    description: 'Spécialiste en analyse de données et rapports',
    color: 'purple',
    permissions: [
      PERMISSIONS.DASHBOARD_VIEW,
      PERMISSIONS.FICHES_VIEW,
      PERMISSIONS.BIOMETRIE_VIEW,
      PERMISSIONS.REPORTS_VIEW,
      PERMISSIONS.REPORTS_CREATE,
      PERMISSIONS.REPORTS_EXPORT,
      PERMISSIONS.IA_VIEW_RESULTS,
      PERMISSIONS.IA_USE_PREDICTION,
      PERMISSIONS.AUDIT_VIEW_OWN,
      PERMISSIONS.ANALYTICS_VIEW,
      PERMISSIONS.EVIDENCE_VIEW,
      PERMISSIONS.NOTIFICATIONS_VIEW,
    ],
  },
}

/**
 * Descriptions lisibles des permissions
 */
export const PERMISSION_LABELS = {
  [PERMISSIONS.DASHBOARD_VIEW]: 'Voir le tableau de bord',
  
  [PERMISSIONS.FICHES_VIEW]: 'Consulter les fiches criminelles',
  [PERMISSIONS.FICHES_CREATE]: 'Créer des fiches criminelles',
  [PERMISSIONS.FICHES_EDIT]: 'Modifier les fiches criminelles',
  [PERMISSIONS.FICHES_DELETE]: 'Supprimer les fiches criminelles',
  
  [PERMISSIONS.BIOMETRIE_VIEW]: 'Consulter les données biométriques',
  [PERMISSIONS.BIOMETRIE_ADD]: 'Ajouter des données biométriques',
  [PERMISSIONS.BIOMETRIE_EDIT]: 'Modifier les données biométriques',
  [PERMISSIONS.BIOMETRIE_DELETE]: 'Supprimer les données biométriques',
  
  [PERMISSIONS.REPORTS_VIEW]: 'Consulter les rapports',
  [PERMISSIONS.REPORTS_CREATE]: 'Générer des rapports',
  [PERMISSIONS.REPORTS_EXPORT]: 'Exporter des rapports',
  
  [PERMISSIONS.NOTIFICATIONS_VIEW]: 'Consulter les notifications',
  
  [PERMISSIONS.IA_VIEW_RESULTS]: 'Voir les résultats IA',
  [PERMISSIONS.IA_USE_RECOGNITION]: 'Utiliser la reconnaissance faciale',
  [PERMISSIONS.IA_USE_PREDICTION]: 'Utiliser l\'analyse prédictive',
  
  [PERMISSIONS.AUDIT_VIEW]: 'Consulter le journal d\'audit',
  [PERMISSIONS.AUDIT_VIEW_OWN]: 'Voir ses propres actions',
  [PERMISSIONS.AUDIT_VIEW_ALL]: 'Voir toutes les actions',
  
  [PERMISSIONS.USERS_VIEW]: 'Consulter les utilisateurs',
  [PERMISSIONS.USERS_CREATE]: 'Créer des utilisateurs',
  [PERMISSIONS.USERS_EDIT]: 'Modifier des utilisateurs',
  [PERMISSIONS.USERS_DELETE]: 'Supprimer des utilisateurs',
  
  [PERMISSIONS.ROLES_VIEW]: 'Consulter les rôles',
  [PERMISSIONS.ROLES_MANAGE]: 'Gérer les rôles et permissions',
}

export default PERMISSIONS

File: backend_gn/utilisateur/permissions.py
"""
Permissions personnalisées pour le système SGIC
Gestion des permissions par rôle
"""

from rest_framework.permissions import BasePermission
from django.contrib.auth import get_user_model

from .models import ADMIN_ROLE_CODES, normalize_role_value

User = get_user_model()

# Cache pour stocker les permissions "gelées" des utilisateurs existants
# Utilisé pour conserver les permissions retirées d'un rôle pour les utilisateurs existants
USER_PERMISSIONS_CACHE = {}


def user_is_app_admin(user):
    """
    Vérifie si l'utilisateur possède un rôle administrateur
    (superuser, staff ou rôle applicatif admin).
    """
    if not user or not getattr(user, 'is_authenticated', False):
        return False

    if user.is_superuser or getattr(user, 'is_staff', False):
        return True

    role_value = normalize_role_value(getattr(user, 'role', None))
    role_code = normalize_role_value(getattr(user, 'role_code', None))
    return role_value in ADMIN_ROLE_CODES or role_code in ADMIN_ROLE_CODES


class IsAdminUserRole(BasePermission):
    """
    Permission simple basée sur le rôle applicatif.
    Autorise uniquement les utilisateurs dont le rôle est administrateur.
    """
    message = "Permission refusée : seuls les administrateurs peuvent effectuer cette action."

    def has_permission(self, request, view):
        return user_is_app_admin(getattr(request, 'user', None))

ROLES_PREDEFINIS = {
    'admin': {
        'permissions': ['*'],
        'description': 'Administrateur applicatif avec tous les droits',
    },
    'Administrateur Système': {
        'permissions': ['*'],  # Tous les droits
        'description': 'Administrateur système avec tous les droits'
    },
    'Enquêteur Principal': {
        'permissions': [
            # Tableau de bord
            'dashboard.view',
            # Fiches criminelles
            'fiches.view',
            'fiches.create',
            'fiches.edit',
            'fiches.delete',
            # Biométrie
            'biometrie.view',
            'biometrie.add',
            'biometrie.edit',
            # Rapports
            'reports.view',
            'reports.create',
            'reports.export',
            # Intelligence Artificielle
            'ia.view_results',
            'ia.use_recognition',
            'ia.use_prediction',
            # Enquêtes
            'investigations.view',
            'investigations.create',
            'investigations.edit',
            'investigations.close',
            # Suspects
            'suspects.view',
            'suspects.create',
            'suspects.edit',
            # Preuves
            'evidence.view',
            'evidence.manage',
            # Analyses
            'analytics.view',
            # Notifications
            'notifications.view',
        ],
        'description': 'Enquêteur principal avec droits de modification'
    },
    'Enquêteur': {
        'permissions': [
            'dashboard.view',
            'fiches.view',
            'fiches.create',
            'fiches.edit',
            'biometrie.view',
            'biometrie.add',
            'reports.view',
            'reports.create',
            'ia.view_results',
            'ia.use_recognition',
            'investigations.view',
            'notifications.view',
        ],
        'description': 'Enquêteur avec droits de base'
    },
    'Enquêteur Junior': {
        'permissions': [
            'dashboard.view',
            'fiches.view',
            'biometrie.view',
            'reports.view',
            'ia.view_results',
            'investigations.view',
            'notifications.view',
        ],
        'description': 'Enquêteur junior avec droits de consultation'
    },
    'Analyste': {
        'permissions': [
            'dashboard.view',
            'fiches.view',
            'biometrie.view',
            'reports.view',
            'reports.create',
            'reports.export',
            'ia.view_results',
            'ia.use_prediction',
            'audit.view_own',
            'analytics.view',
            'evidence.view',
            'notifications.view',
        ],
        'description': 'Analyste avec droits d\'analyse'
    },
    'Observateur': {
        'permissions': [
            'dashboard.view',
            'fiches.view',
            'biometrie.view',
            'reports.view',
            'reports.export',
            'notifications.view',
            'ia.view_results',
            'investigations.view',
            'suspects.view',
            'evidence.view',
        ],
        'description': 'Observateur avec droits de consultation uniquement'
    },
}

PERMISSIONS = {
    'view_utilisateur': 'Voir les utilisateurs',
    'add_utilisateur': 'Créer des utilisateurs',
    'change_utilisateur': 'Modifier les utilisateurs',
    'delete_utilisateur': 'Supprimer les utilisateurs',
    'view_criminel': 'Voir les fiches criminelles',
    'add_criminel': 'Créer des fiches criminelles',
    'change_criminel': 'Modifier les fiches criminelles',
    'delete_criminel': 'Supprimer les fiches criminelles',
    'view_rapport': 'Voir les rapports',
    'add_rapport': 'Créer des rapports',
    'change_rapport': 'Modifier les rapports',
    'delete_rapport': 'Supprimer les rapports',
    
    'dashboard.view': 'Voir le tableau de bord',
    'fiches.view': 'Consulter les fiches criminelles',
    'fiches.create': 'Créer des fiches criminelles',
    'fiches.edit': 'Modifier les fiches criminelles',
    'fiches.delete': 'Supprimer les fiches criminelles',
    'biometrie.view': 'Consulter les données biométriques',
    'biometrie.add': 'Ajouter des données biométriques',
    'biometrie.edit': 'Modifier les données biométriques',
    'biometrie.delete': 'Supprimer les données biométriques',
    'reports.view': 'Consulter les rapports',
    'reports.create': 'Générer des rapports',
    'reports.export': 'Exporter des rapports',
    'notifications.view': 'Consulter les notifications',
    'ia.view_results': 'Voir les résultats IA',
    'ia.use_recognition': 'Utiliser la reconnaissance faciale',
    'ia.use_prediction': 'Utiliser l\'analyse prédictive',
    'audit.view': 'Consulter le journal d\'audit',
    'audit.view_own': 'Voir ses propres actions',
    'audit.view_all': 'Voir toutes les actions',
    'users.view': 'Consulter les utilisateurs',
    'users.create': 'Créer des utilisateurs',
    'users.edit': 'Modifier des utilisateurs',
    'users.delete': 'Supprimer des utilisateurs',
    'roles.view': 'Consulter les rôles',
    'roles.manage': 'Gérer les rôles et permissions',
    'investigations.view': 'Voir les enquêtes',
    'investigations.create': 'Créer des enquêtes',
    'investigations.edit': 'Modifier les enquêtes',
    'investigations.close': 'Clôturer les enquêtes',
    'suspects.view': 'Voir les suspects',
    'suspects.create': 'Créer des suspects',
    'suspects.edit': 'Modifier les suspects',
    'evidence.view': 'Voir les preuves',
    'evidence.manage': 'Gérer les preuves',
    'analytics.view': 'Voir les analyses',
}


def get_user_permissions(user):
    """
    Récupère les permissions d'un utilisateur selon son rôle
    Si l'utilisateur a des permissions "gelées" (stockées dans le cache),
    elles sont fusionnées avec les permissions actuelles du rôle
    """
    if not user or not user.is_authenticated:
        return []
    
    # Administrateur Système a tous les droits
    if user_is_app_admin(user):
        return ['*']
    
    # Récupérer les permissions selon le rôle
    role_value = (getattr(user, 'role', '') or '').strip()
    role_permissions = []
    if role_value and role_value in ROLES_PREDEFINIS:
        role_permissions = ROLES_PREDEFINIS[role_value].get('permissions', [])
    
    user_id = getattr(user, 'id', None)
    if user_id and user_id in USER_PERMISSIONS_CACHE:
        # Fusionner les permissions du rôle avec les permissions gelées
        frozen_permissions = USER_PERMISSIONS_CACHE[user_id]
        # Union des permissions : on garde les permissions gelées + les nouvelles permissions du rôle
        merged_permissions = set(role_permissions) | frozen_permissions
        return list(merged_permissions)
    
    return role_permissions


def freeze_user_permissions(user, permissions_to_freeze):
    """
    "Gèle" (conserve) certaines permissions pour un utilisateur
    Utilisé quand on retire une permission d'un rôle pour garder cette permission pour l'utilisateur existant
    """
    if not user or not hasattr(user, 'id'):
        return
    
    user_id = user.id
    if user_id not in USER_PERMISSIONS_CACHE:
        USER_PERMISSIONS_CACHE[user_id] = set()
    
    # Ajouter les permissions à conserver
    USER_PERMISSIONS_CACHE[user_id].update(permissions_to_freeze)


def clear_user_permissions_cache(user_id=None):
    """
    Nettoie le cache des permissions utilisateur
    Si user_id est fourni, nettoie uniquement pour cet utilisateur
    """
    global USER_PERMISSIONS_CACHE
    if user_id:
        USER_PERMISSIONS_CACHE.pop(user_id, None)
    else:
        USER_PERMISSIONS_CACHE.clear()


def has_permission(user, permission_code):
    """
    Vérifie si un utilisateur a une permission spécifique
    """
    permissions = get_user_permissions(user)
    
    # Si l'utilisateur a tous les droits
    if '*' in permissions:
        return True
    
    # Vérifier la permission spécifique
    return permission_code in permissions


def get_role_permissions(role_name):
    """
    Récupère les permissions d'un rôle
    """
    return ROLES_PREDEFINIS.get(role_name, {}).get('permissions', [])


def has_role_permission(user, permission_code):
    """
    Vérifie si un utilisateur a une permission spécifique liée aux rôles
    """
    if not user or not user.is_authenticated:
        return False
    
    # Superuser et admin système ont toutes les permissions
    if user_is_app_admin(user):
        return True
    
    # Vérifier les permissions selon le rôle
    user_permissions = get_user_permissions(user)
    if '*' in user_permissions:
        return True
    
    return permission_code in user_permissions
